apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -37,25 +41,59 @@ interface DanceAiProps {
   onClose: () => void;
 }
 
+// Steps in our generating process have a step and a substep.
+type GeneratingStep = [number, number];
+
+// Adapted from https://overreacted.io/making-setinterval-declarative-with-react-hooks/
+function useInterval(callback: () => void, delay: number | undefined) {
+  const savedCallback = useRef<() => void>();
+
+  // Remember the latest callback.
+  useEffect(() => {
+    savedCallback.current = callback;
+  }, [callback]);
+
+  // Set up the interval.
+  useEffect(() => {
+    function tick() {
+      if (savedCallback.current) {
+        savedCallback.current();
+      }
+    }
+    if (delay !== undefined) {
+      const id = setInterval(tick, delay);
+      return () => clearInterval(id);
+    }
+  }, [delay]);
+}
+
 const getImageUrl = (id: string) => {
   return `/blockly/media/dance/ai/emoji/${id}.svg`;
 };
 
 const DanceAiModal: React.FunctionComponent<DanceAiProps> = ({onClose}) => {
   const dispatch = useAppDispatch();
 
+  // How many emojis are to be selected.
   const SLOT_COUNT = 3;
 
+  // How many low-scoring results we show before the chosen one.
+  const BAD_GENERATED_RESULTS_COUNT = 4;
+
+  // How many substeps for each step in the generating process.
+  const GENERATING_SUBSTEP_COUNT = 2;
+
+  // How long we spend in each substep in the generating process.
+  const GENERATION_SUBSTEP_DURATION = 1000;
+
+  const generatedEffects = useRef<GeneratedEffect[]>([]);
+
   const [mode, setMode] = useState(Mode.SELECT_INPUTS);
   const [currentInputSlot, setCurrentInputSlot] = useState(0);
   const [inputs, setInputs] = useState<string[]>([]);
-  const [resultJson, setResultJson] = useState<string>('');
-  const [generatingNodesDone, setGeneratingNodesDone] =
-    useState<boolean>(false);
   const [processingDone, setProcessingDone] = useState<boolean>(false);
-  const [generatingDone, setGeneratingDone] = useState<boolean>(false);
-  const [typingDone, setTypingDone] = useState<boolean>(false);
-  const [showPreview, setShowPreview] = useState<boolean>(false);
+  const [generatingStep, setGeneratingStep] = useState<GeneratingStep>([0, 0]);
+  const [currentToggle, setCurrentToggle] = useState<string>('ai-block');
```

Comment:
explicit type for this (`'ai-block' | 'code'`)?

Reply:
Good call.